[Merged by Bors] - feat(order/complete_well_founded, data/finset/lattice): introduce compact elements of a complete lattice and characterize compact lattices as those with all elements compact

cemulate opened the pull request:
Provide a definition of compact elements. Prove the equivalence of two characterizations of compact elements. Add "all elements are compact" to the equivalent characterizations of compact lattices. Introduce an auxiliary lemma about finite sups and directed sets.

<!--
A reference for the two equivalent definitions of compact element can be found [here](https://ncatlab.org/nlab/show/compact+element+in+a+lattice)
-->

cemulate committed:
auxiliary lemma about finite sups and directed sets

cemulate committed:
define compact elements; add "all elements are compact" to the characterizations of compact lattices

ocfnash commented on src/data/finset/lattice.lean:
```diff
@@ -119,6 +119,34 @@ begin
     { rw [finset.sup_insert, id.def], exact h h_subset.1 (h₁ htne h_subset.2), }, },
 end
 
+lemma sup_le_of_le_directed {α : Type*} [complete_lattice α] (s : set α)
```
```suggestion
lemma sup_le_of_le_directed {α : Type*} [semilattice_sup_bot α] (s : set α)
```

ocfnash commented on src/data/finset/lattice.lean:
```diff
@@ -119,6 +119,34 @@ begin
+    by_cases hs : s.nonempty,
+    { intros _ as_below,
+      have incs : ↑s ⊆ ↑(insert a s), by { rw finset.coe_subset, apply finset.subset_insert, },
+      rcases (ih hs (set.subset.trans incs as_below)) with ⟨x, ⟨hxs, hsx_sup⟩⟩,
```
```suggestion
      obtain ⟨x, ⟨hxs, hsx_sup⟩⟩ := ih hs (set.subset.trans incs as_below),
```
(and likewise for the other two `rcases`)

ocfnash commented on src/data/finset/lattice.lean:
```diff
@@ -119,6 +119,34 @@ begin
 
+lemma sup_le_of_le_directed {α : Type*} [complete_lattice α] (s : set α)
+  (hdir : directed_on (≤) s) (t : finset α) :
+  t.nonempty → (∀ x ∈ t, ∃ y ∈ s, x ≤ y) → ∃ x, x ∈ s ∧ t.sup id ≤ x :=
```
I think it might be preferable to assume `s.nonempty` and drop the `t.nonempty` condition.

cemulate replied:
I like that. That simplifies this lemma a lot and `s.nonempty` is the case in the only place I apply it.

ocfnash commented on src/order/complete_well_founded.lean:
```diff
@@ -49,6 +49,61 @@ same `Sup`. -/
+/-- An equivalent characterization of a compact element k is that any directed
+that Sups above k must get above k at some point in the set -/
+theorem is_compact_element_iff_directed_Sup_finite (k : α) :
+  is_compact_element k ↔ ∀ s : set α, s.nonempty → directed_on (≤) s → k ≤ Sup s → ∃ x : α, x ∈ s ∧ k ≤ x :=
```
If I'm reading [this page](https://encyclopediaofmath.org/wiki/Compact_lattice_element) correctly, the RHS of your iff is the condition of being a finite element.

So should we introduce:
```lean
def is_finite_element {α : Type*} [complete_lattice α] (k : α) :=
∀ s : set α, s.nonempty → directed_on (≤) s → k ≤ Sup s → ∃ x : α, x ∈ s ∧ k ≤ x
```
above and then use that here?

cemulate replied:
Ah yes, those concepts can be distinct, but are equivalent for complete lattices. Given that, I think it'd be nice to define finite element and show equivalence in the complete lattice case.

ocfnash commented on src/order/complete_well_founded.lean:
```diff
@@ -49,6 +49,61 @@ same `Sup`. -/
+    let S : set α := { x | ∃ t : finset α, ↑t ⊆ s ∧ x = t.sup id },
+    -- S is directed, nonempty, and still sups above k.
+    have dir_US : directed_on (≤) S,
+    { intros x hx y hy,
```
```suggestion
    { rintros x ⟨c, hc⟩ y ⟨d, hd⟩,
```
and then you can drop the next two lines.

ocfnash commented on src/order/complete_well_founded.lean:
```diff
@@ -49,6 +49,61 @@ same `Sup`. -/
+    -- Any nonempty directed set certainly sups above ⊥
+    { rintros _ _ ⟨x, hx⟩ _ _, use x, by simp[hx, hbot], },
+    { intros hk s hne hdir hsup,
+      cases (hk s hsup) with t ht,
```
I slightly prefer
```suggestion
      obtain ⟨t,  ht⟩ := hk s hsup,
```
and elsewhere instead of the various `cases`.

ocfnash commented on src/order/complete_well_founded.lean:
```diff
@@ -109,9 +188,14 @@ lemma is_sup_closed_compact_iff_well_founded :
 alias is_Sup_finite_compact_iff_is_sup_closed_compact ↔
       _ is_sup_closed_compact.is_Sup_finite_compact
 alias is_sup_closed_compact_iff_well_founded ↔ _ well_founded.is_sup_closed_compact
+alias is_Sup_finite_compact_iff_all_elements_compact ↔ _ all_elements_compact.is_Sup_finite_compact
```
```suggestion
```
As it stands, this line creates an alias for the `←` direction of `is_Sup_finite_compact_iff_all_elements_compact` and calls it `all_elements_compact.is_Sup_finite_compact`. But `all_elements_compact.is_Sup_finite_compact` already exists so this line does nothing!

Now that we have 4 equivalent conditions, we have 12 possible implications between them. I wonder if we should only create aliases for `well_founded.foo` for the three values of `foo`?

cemulate replied:
I'm a bit unsure what the utility of these aliases is. In any case where they were needed, one could simply use `{iff lemma}.(mp|mpr)`, no?

ocfnash replied:
It might be worth seeking @urkud 's opinion on the aliases.

ocfnash commented on src/order/complete_well_founded.lean:
```diff
@@ -109,9 +188,14 @@ lemma is_sup_closed_compact_iff_well_founded :
   is_sup_closed_compact α ↔ well_founded ((>) : α → α → Prop) :=
 (well_founded_characterisations α).out 2 0
 
+lemma is_Sup_finite_compact_iff_all_elements_compact :
```
Given that we essentially prove this above (via `all_elements_compact.is_Sup_finite_compact` and `is_Sup_finite_compact.all_elements_compact`) I think it might be better just to state and prove this above and not go via the `tfae` machinery.

cemulate replied:
Yes, since this is already an iff between 2 & 4, it seems like a good idea to just establish the iff above; I'll do that.

cemulate committed:
Clean up sup_le_of_le_directed

Replace rcases with obtains; simplify hypothesis making the proof much shorter.

cemulate committed:
Squeeze a simp

cemulate committed:
Fix application of finset lemma

cemulate committed:
Rephrase compact element equivalent definitions in terms of "compact" and "finite"

cemulate committed:
squeeze a simpa

cemulate committed:
Clean up by replacing rcases w/ obtains

cemulate committed:
Combine equivalence of is_Sup_finite_compact and all_elements_compact in a single lemma.

Use the iff in the TFAE

cemulate committed:
Another rcases -> obtain

ocfnash commented on src/order/complete_well_founded.lean:
```diff
@@ -88,12 +147,31 @@ begin
+begin
+  split,
+  { intros h k s hs,
+    rcases (h s) with ⟨t, ⟨hts, htsup⟩⟩,
```
```suggestion
    obtain ⟨t, ⟨hts, htsup⟩⟩ := h s,
```

ocfnash commented on src/order/complete_well_founded.lean:
```diff
@@ -88,12 +147,31 @@ begin
+
 lemma well_founded_characterisations :
-  tfae [well_founded ((>) : α → α → Prop), is_Sup_finite_compact α, is_sup_closed_compact α] :=
+  tfae [well_founded ((>) : α → α → Prop), is_Sup_finite_compact α, is_sup_closed_compact α, ∀ k : α, is_compact_element k] :=
```
```suggestion
  tfae [well_founded ((>) : α → α → Prop),
        is_Sup_finite_compact α,
        is_sup_closed_compact α,
        ∀ k : α, is_compact_element k] :=
```

cemulate committed:
rcases -> obtain

Co-authored-by: Oliver Nash <github@olivernash.org>

cemulate committed:
Better layout on long TFAE

Co-authored-by: Oliver Nash <github@olivernash.org>

awainverse commented on src/order/complete_well_founded.lean:
```diff
@@ -49,6 +49,65 @@ same `Sup`. -/
+
+/-- An element k of a lattice is said to be finite if any directed
+set that Sups above k already got above k at some point in the set. -/
+def is_finite_element {α : Type*} [complete_lattice α] (k : α) :=
```
I know this is going contrary to @ocfnash 's advice, but I don't think there should be separate names for `is_compact_element` and `is_finite_element`. At least, if both are only defined for elements of a `complete_lattice`, where they are equivalent. If there's a natural way to define these predicates outside the `complete_lattice` context, where they're not equivalent, then do it, but otherwise, I think we're just creating two names for the same thing, which complicates the API.

ocfnash replied:
I still lean in favour of the two names but I'm happy to defer to @awainverse on this point!

cemulate replied:
The source that [this page](https://encyclopediaofmath.org/wiki/Compact_lattice_element) cites defines them both for complete lattices, but remarks that the "finite" defn. makes sense for lattices where only Sups of directed sets necessarily exist. But, it's mostly a moot point and I think the only reason there are two definitions is that "finite" came from computer science.

If we don't use both names, then how would you recommend naming the iff lemma? Something like the following?
```
theorem is_compact_element_iff_le_of_directed_Sup_le (k : α) :
  is_compact_element k ↔ ∀ s : set α, s.nonempty → directed_on (≤) s → k ≤ Sup s → ∃ x : α, x ∈ s ∧ k ≤ x
```

awainverse replied:
That seems fine to me.

awainverse replied:
You could also maybe add a docstring saying that this property also defines finite elements, so that if someone does have a use for that idea separately, they know where to look.

awainverse reviewed (commented):
I've left a comment on the naming of `is_finite_element`, but otherwise, this looks good to me. Maybe we could also consider renaming the file, as the gist of this file to me is about compactness in lattices.
Also, lint-style is upset, but I'm not sure why, because it seems the long lines have been fixed.

ocfnash commented on src/order/complete_well_founded.lean:
```diff
@@ -49,6 +49,65 @@ same `Sup`. -/
+      have tne : t.nonempty,
+      { by_contradiction n,
+        rw [finset.nonempty_iff_ne_empty, not_not] at n,
+        simp only [n, true_and, set.empty_subset, finset.coe_empty, finset.sup_empty, le_bot_iff] at ht,
```
Style linter says:
`src/order/complete_well_founded.lean : line 77 : ERR_LIN : Line has more than 100 characters`

ocfnash commented on src/order/complete_well_founded.lean:
```diff
@@ -49,6 +49,65 @@ same `Sup`. -/
+        rw [finset.nonempty_iff_ne_empty, not_not] at n,
+        simp only [n, true_and, set.empty_subset, finset.coe_empty, finset.sup_empty, le_bot_iff] at ht,
+        exact absurd ht hbot, },
+      -- certainly ↑t ⊆ s, so every element of t is below something in s, to apply finset_sup_of_directed_on
```
Style linter says:
`src/order/complete_well_founded.lean : line 79 : ERR_LIN : Line has more than 100 characters`

ocfnash commented on src/order/complete_well_founded.lean:
```diff
@@ -49,6 +49,65 @@ same `Sup`. -/
+    have Sne : S.nonempty,
+    { suffices : ⊥ ∈ S, from set.nonempty_of_mem this,
+      use ∅,
+      simp only [set.empty_subset, finset.coe_empty, finset.sup_empty, eq_self_iff_true, and_self], },
```
Style linter says:
`src/order/complete_well_founded.lean : line 104 : ERR_LIN : Line has more than 100 characters`

ocfnash reviewed (commented):
I've highlighted the lines that still contain > 100 characters. You'll need to break these to appease the CI.

cemulate committed:
Merge branch 'master' into complete-lattice-compact-elements

cemulate committed:
Remove mention of "finite elements", just keep "compact"

cemulate committed:
Add definition & theorem about compactly generated lattices

cemulate committed:
Update documentation

cemulate committed:
Rename complete_well_founded -> compactly_generated, fix reference

cemulate committed:
Fix long lines

awainverse reviewed (approved):
LGTM

bryangingechen commented on src/order/compactly_generated.lean:
```diff
@@ -0,0 +1,217 @@
+ * `is_sup_closed_compact`
+ * `is_Sup_finite_compact`
+ * `is_compact_element`
+ * `is_compactly_generated`
```
```suggestion
 * `complete_lattice.is_sup_closed_compact`
 * `complete_lattice.is_Sup_finite_compact`
 * `complete_lattice.is_compact_element`
 * `complete_lattice.is_compactly_generated`
```

cemulate replied:
Personally, in informal English, I would say "joins above" or "join gets above", but I don't really know if this is common. I'm perfectly happy to accept these wording changes though.

bryangingechen commented on src/order/compactly_generated.lean:
```diff
@@ -0,0 +1,217 @@
+ * `well_founded (>)`
+ * `is_sup_closed_compact`
+ * `is_Sup_finite_compact`
+ * `∀ k, is_compact_element k`
```
```suggestion
 * `complete_lattice.is_sup_closed_compact`
 * `complete_lattice.is_Sup_finite_compact`
 * `∀ k, complete_lattice.is_compact_element k`
```

bryangingechen commented on src/order/compactly_generated.lean:
```diff
@@ -0,0 +1,217 @@
+ * `well_founded.is_Sup_finite_compact`
+ * `is_Sup_finite_compact.is_sup_closed_compact`
+ * `is_sup_closed_compact.well_founded`
+ * `is_Sup_finite_compact_iff_all_elements_compact`
```
```suggestion
 * `complete_lattice.well_founded.is_Sup_finite_compact`
 * `complete_lattice.is_Sup_finite_compact.is_sup_closed_compact`
 * `complete_lattice.is_sup_closed_compact.well_founded`
 * `complete_lattice.is_Sup_finite_compact_iff_all_elements_compact`
```

bryangingechen commented on src/order/compactly_generated.lean:
```diff
@@ -0,0 +1,217 @@
+ * `is_sup_closed_compact.well_founded`
+ * `is_Sup_finite_compact_iff_all_elements_compact`
+
+ We also show well-founded lattices are compactly generated.
```
```suggestion
 We also show well-founded lattices are compactly generated
 (`complete_lattice.compactly_generated_of_well_founded`).
```

bryangingechen commented on src/order/compactly_generated.lean:
```diff
@@ -0,0 +1,217 @@
+∀ (s : set α), ∃ (t : finset α), ↑t ⊆ s ∧ Sup s = t.sup id
+
+/-- An element k of a complete lattice is said to be compact if any set that Sups
+above k has a finite subset that Sups above k -/
```
I moved some text from the docstring of `is_compact_element_iff_le_of_directed_Sup_le` here:
```suggestion
/-- An element `k` of a complete lattice is said to be compact if any set with `Sup`
above `k` has a finite subset with `Sup` above `k`.  Such an element is also called
"finite" or "S-compact". -/
```

bryangingechen commented on src/order/compactly_generated.lean:
```diff
@@ -0,0 +1,217 @@
+∀ s : set α, k ≤ Sup s → ∃ t : finset α, ↑t ⊆ s ∧ k ≤ t.sup id
+
+/-- A complete lattice is said to be compactly generated if any
+element is the Sup of compact elements-/
```
```suggestion
element is the `Sup` of compact elements. -/
```

bryangingechen commented on src/order/compactly_generated.lean:
```diff
@@ -0,0 +1,217 @@
+
+/-- Equivalently, an element k compact when any directed set that Sups above
+k already got above k at some point in the set. Such an element is also called
+"finite" or "S-compact" -/
```
I removed "Equivalently" since it won't be too meaningful in a hover pop-up.
```suggestion
/-- An element `k` is compact if and only if any directed set with `Sup` above
`k` already got above `k` at some point in the set. -/
```

bryangingechen commented on src/order/compactly_generated.lean:
```diff
@@ -0,0 +1,217 @@
+  classical,
+  split,
+  { by_cases hbot : k = ⊥,
+    -- Any nonempty directed set certainly sups above ⊥
```
```suggestion
    -- Any nonempty directed set certainly has sup above ⊥
```

bryangingechen commented on src/order/compactly_generated.lean:
```diff
@@ -0,0 +1,217 @@
+    { rintros _ _ ⟨x, hx⟩ _ _, use x, by simp only [hx, hbot, bot_le, and_self], },
+    { intros hk s hne hdir hsup,
+      obtain ⟨t, ht⟩ := hk s hsup,
+      -- If t were empty, it would sup to ⊥, which is not above k ≠ ⊥.
```
```suggestion
      -- If t were empty, its sup would be ⊥, which is not above k ≠ ⊥.
```

bryangingechen commented on src/order/compactly_generated.lean:
```diff
@@ -0,0 +1,217 @@
+  { intros hk s hsup,
+    -- Consider the set of finite joins of elements of the (plain) set s.
+    let S : set α := { x | ∃ t : finset α, ↑t ⊆ s ∧ x = t.sup id },
+    -- S is directed, nonempty, and still sups above k.
```
```suggestion
    -- S is directed, nonempty, and still has sup above k.
```

bryangingechen reviewed (commented):
Is "sups above" common terminology? I couldn't find any sources that used this.

cemulate committed:
Documentation and wording improvements

Co-authored-by: Bryan Gin-ge Chen <bryangingechen@gmail.com>

bryangingechen reviewed (approved):
Thanks!
bors r+
